sweepgen.1
.TH sweepgen 1 "20 Feb 1998" "Release 2" "Linux System Manual"
.SH NAME
.I sweepgen \- an Ncurses based sweep generator program
.SH SYNOPSIS
.IP sweepgen\ [options]
.SH DESCRIPTION
.I sweepgen
is a simple sweep signal generator program, with an Ncurses based user 
interface, that can digitally generate waveforms on the LINUX 
.B /dev/dsp
device.
8 or 16 bit samples can be generated depending on the hardware.
.PP
.I sweepgen
in 
.B MONO
mode outputs the generated swept waveform. In 
.B STEREO
mode the swept waveform is output on one channel, and the sweeping waveform
is output on the second channel.
.PP
The sweeping frequency is by default
specified as an integer number of Hertz. Fractional Hertz frequencies,
specifiable to 0.1Hz or 0.01Hz resolution, are
supported by use of the 
.I -res
command line option, or the
.I resolution
parameter in the configuration file(s). However
.I be\ warned
at 0.1Hz resolution the basic waveform sample buffers generated 
are each 10 times
(and at 0.01Hz resolution 100 times) as big as the samplerate. It typically
requires 5.5Mbytes of memory to run at 0.1Hz resolution, 16bit 32000
samples/sec. and 55Mbytes of memory to run at 0.01Hz resolution. Because of the
large buffer sizes, the initial waveform calculation time can also be lengthy.
Remember also that the waveforms are re-calculated whenever the playing
parameters, 8/16bit, mono/stereo, samplerate are changed.

.PP
The swept waveform frequencies are specified either as a lower and upper
frequency, or as a centre frequency with a frequency variation.
.IP 1000\ 500
would signify a range of 1000-500 Hz to 1000+500 Hz, i.e. 500 to 1500 Hz
.PP
Of course, only frequencies less than half the samplerate
(number of samples/sec) can be generated. Although this is not checked. 
.PP 
The waveforms that can be used as either the sweeping or swept signals are:
.IP sine
A standard sine wave
.IP square
a standard square wave with a 50% mark space ratio
.IP triangle
a linear rise from 0 to peak, thru' 0 to negative peak, and back to 0
.IP sawtooth
a ramp waveform with 'infinitely' fast flyback (:-) An ideal oscilloscope
timebase signal.
.IP noise
This is weak. All it consists of is one second of pseudo-randomly generated
samples, played repeatedly. I'd love to do proper white/pink noise,
but I don't know enough, and I don't think the structure of the program
is conducive to accurate noise generation.
.IP pulse
A square waveform where the mark/space ratio (as a percentage)
is 10% (mark/space ratio of 1:9).
.PP
A lot of thought has gone into the algorithms for generating the waveforms.
I believe the sin/cos wave to be very pure (modulo your sound card :-), but
I don't have access to a THD meter to measure it. 
.PP
At 1Hz resolution,
.I sweepgen
generates one seconds worth of 1 Hz samples at the specified samplerate, for
each waveform, and generates sweeping frequency F by circularly sampling
every Fth sample. These samples are scaled to fit the swept frequency range
and are used to sample the swept waveform to generate the swept signal.
Each buffer fragment is generated for the parameter(s)
set at that moment. By default, buffer fragment sizes are set so that aprox. 10
fragments/sec are generated. Changing a generation parameter, e.g. waveform,
frequency, will impact the next buffer fragment generated, and hence
changes appear to be almost immediate.
.PP
If your sounds periodically 'breaks' up with clicks or breaks,
it is usually a sign
that siggen is not being scheduled sufficiently often. Either up the
priority (see 
.I nice
et al.), kill off other processes, get a faster processor, or increase the
number of audio buffer fragments that siggen uses. This last will make
siggen respond more sluggishly to changes in generation parameters.
.I syslogd
and
.I crond
are two processes that I've found useful to kill off - YMMV.
.IP Defaults
output to /dev/dsp, 22050 samples/sec, stereo if stereo card else mono,
16 bit samples if possible, else 8 bit. 
.SH CONFIGURATION\ FILES
.PP
Three possible configuration files can be used: a LOCAL config file (usually
in current directory), a HOME config file in user's $HOME directory and a
GLOBAL config file.
.PP
All the siggen suite of programs are compiled with the names of the config
files built in. By default the configuration files are:
.IP ./.siggen.conf
is the LOCAL config file.
.IP $HOME/.siggen.conf
is the HOME config file.
.IP /etc/siggen.conf
is the GLOBAL config file.
.IP siggen\ -h
will indicate which config files will be searched for.
.PP
The config files do not have to exist. If they exist and are readable by the
program they are used, otherwise they are simply ignored.
.PP
The config files are always searched for configuration values in the order
LOCAL, HOME, GLOBAL. This allows a scheme where the sysadmin sets up default
config values in the GLOBAL config file, but allows a user to set some or
all different values in their own HOME config file, and to set yet more
specific values when run from a particular directory.
.PP
If no configuration files exist, the program provides builtin
default values, and most of these values can be set
by appropriate command line switches and flags.
.PP
See siggen.conf(5) for details of the configuration files.
.PP
.I sweepgen
looks for configuration values BUFFERSPERSEC, CHANNELS, DACFILE, FRAGMENTS,
RESOLUTION, SAMPLERATE, SAMPLESIZE, VERBOSE, VI_KEYS.
.IP BUFFERSPERSEC
The aprox. number of sound buffer fragments to play every second
(Sound buffersize is always a power of 2).
.IP CHANNELS
sets the number of channels, mono or stereo.
.IP DACFILE
allows the name of the DAC/DSP/PCM device to be changed from /dev/dsp
.IP FRAGMENTS
The number of Audio Buffers to configure in the driver.
.IP RESOLUTION
The minimum change possible to the frequency setting. Only 3 values allowed:
1Hz , 0.1Hz or 0.01Hz
.IP SAMPLERATE
sets the number of samples/sec for the DAC device
.IP SAMPLESIZE
sets whether 8 or 16 bit samples to be generated
.IP VERBOSE
sets whether or not to run in verbose mode.
.IP VI_KEYS
if set then the VI cursor moving keys "HJKL" are enabled

.SH OPTIONS
.IP -h
display usage and help info
.IP -v
be verbose
.IP -s\ samples
generate with samplerate of samples/sec
.IP -8|-16\ or\ -b\ 8|16
force 8 bit or 16 bit mode.
.IP -1|-2
mono or stereo mode
.IP -res\ n
set resolution of frequency generation. Valid values are: 1Hz, 0.l1Hz or 0.01Hz
.SH EXAMPLES
.SH
.SH FILES
.SH
.SH SEE ALSO
swgen, sgen, siggen, tones
.SH BUGS
.SH
.SH COPYING
.I Copyright\ 1995-2008\ Jim\ Jackson
.PP
The software described by this manual is covered by the GNU General
Public License, Version 2, June 1991, issued by :
.IP
Free Software Foundation, Inc.,
.br
675 Mass Ave,
.br
Cambridge, MA 02139, USA
.PP
Permission is granted to make and distribute verbatim copies of
this manual provided the copyright notice and this permission notice
are preserved on all copies.
.PP
Permission is granted to copy and distribute modified versions of this
manual under the conditions for verbatim copying, provided that the
entire resulting derived work is distributed under the terms of a
permission notice identical to this one.
.PP
Permission is granted to copy and distribute translations of this
manual into another language, under the above conditions for modified
versions, except that this permission notice may be included in
translation instead of in the original English.
.SH AUTHOR
.I Jim Jackson
.br
.sp
.I Email: jj@franjam.org.uk
.br


CHANGES

siggen - a package of waveform generation programs for linux
------------------------------------------------------------

Version 2.3.+

  siggen
  
    Tried adding loadable waveforms - came to an impasse. I will revisit
    this.
    Added a "tracking" feature suggested by Gart Reese. 't' or 'T'
    while in the waveform, frequency or gain fields will cause both
    similar fields of each channel to track, i.e. have the same value.
    A "tracking" message shows which fields are currently tracking.
    Repeating 't' or 'T' cancels the tracking of the fields.
    
  fsynth
    
    A curses based "fourier" synthesis program. Allows a mixture of
    a fundamental sine wave and varying proportions of it's harmonics
    to be mixed to create complex waveforms.
    
  tones

    Added loadable waveforms. Tones can load a waveform from a WAV file
    that contains exactly one seconds worth of samples, at the same samplerate
    as tones is using and same format etc, and it will "play" from the these
    samples. The waveform name is the basename of the wav file with directory
    and any suffix removed. To work correctly the samples should be for a 1Hz
    wave. If its for say a 2Hz wave then the frequencies generated will all
    be double for that waveform.
    Added recursive command file reading.
    Attenuation/gain levels added, thanks to "Mark E. Shoulson"
    <mark@kli.org> for a patch, feedback and comments. Fixed faulty mixing
    code for more than 2 channels.

  All curses progs

    The ability to "Lock" input field to freeze values. Locked fields
    are skipped while moving around the screen. All locked fields can 
    be unlocked in one go to reset. '!' or 'L' locks the current field,
    'U' unlocks all locked fields. This was the result of misunderstanding
    an idea from Garst!
    
    Also added ability to redraw screen by pressing Ctrl-L (^L), useful
    if screen gets garbled.

    Bug fix in scfio.c, thanks to "Garst R. Reese" <reese@isn.net>
    for use egcs and finding the bug - I think gcc was too forgiving.

  All progs
    
    Added configuration file(s) support. 

Version 2.2

  tones, sgen, swgen

    Added ability to compile on machines without sound support,
    so that WAV and raw data files can be created.

    Fixed so can be compiled under SUNOS on sparc machines. This meant
    fixing the implicit little endian nature of the raw and wav file writing
    code.
   
    New config.h file added with user configurable options gathered together
    in one place, for global options.

  tones, sgen, swgen

    Added the -x10 and -x100 options, which when in use allow the
    generation of freq. to the nearest 0.1Hz or 0.01Hz by forcing
    the specification of all frequencies to be 10 or 100 times
    too big. 

    e.g.  swgen -x10 -2 -v 2 2000 20000

          causes stereo samples for a 0.2Hz sawtooth ramp sweeping
          a sin wave from 200Hz to 2000Hz.

Version 2.1

   Fixed compilation problem of incompatible extern declaration
     of sys_errlist . extern wasn't necessary so deleted it.

Version 2

   Added swgen - a swept frequency generator.

   Added sweepgen - a curses based swept freq. generator

   Changed siggen to play continuously - changes to waveform, freq,
     gain etc take effect (almost) immediately. Downside is that
     this is a lot more processor intensive. Fine, just, on my 486DX2/66.

   Extended the tones program, extensively.

   Added dtmf program which can be used as tone phone dialer.

   Fixed mixed up description of sawtooth/triangle waves in sgen manual.

Version 1.1  released xx March 1997

   Fixed a small initialise bug in sgen - reported by 
      Michael Meifert <mime@dk3hg.hanse.de>  
      (Michael this email address doesn't work)

   Fixed a bug in siggen that prevented PULSE waveforms being generated
      
   Added the tones program following a suggestion by Michael who
      wanted a program to generate tones to control a radio repeater (I think).


Version 1    released  6 March 1997

   contained programs sgen, siggen, smix, soundinfo




sweepgen.c
/* sweepgen.c
 * Ncurses based Sweep generator, 16/8 bit, 2 channels if stereo card,
 * Emulates an audio sweep signal generator
 * both the swept and sweeping frequencies can be
 *   sine, square, triangle, sawtooth, pulse, noise waveforms
 * 
 * Linux Version
 */

/*
 * Copyright (C) 1997-2008 Jim Jackson                    jj@franjam.org.uk
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program - see the file COPYING; if not, write to 
 *  the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, 
 *  MA 02139, USA.
 */

/*
 * sweepgen :
 * 
 * 1 secs worth of 1 Hz samples for each waveform are generated, and the
 * sweeping frequency F is generated by picking out every Fth sample
 * The swept freq, f is generated by using the samples from F to 
 * increment a selection pointer out of the swept waveform buffer.
 * The selected samples are stored in a playbuff that is fed to the
 * DSP when full.
 * 
 * 
 * History:
 *  2002    V1.2   Added fractional frequency settings, can be setup to
 *                 generate freq. in 1Hz, 0.1Hz or 0.01Hz resolution
 *
 *  03Nov97 V1.0  Started based on siggen 1.5
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <errno.h>
#include <ctype.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/soundcard.h>
#include <math.h>
#include "sweepgen.h"

#define MAXPRM 32
#define chkword(a,b) ((n>=a)&&(strncmp(p,b,n)==0))

int vflg,dflg,vikeys;

int resolution;                  /* Freq. resolution for signal generation
                                  * 1 for 1Hz res ; 10 for 0.1Hz res;
                                  * 100 for 0.01Hz res                   */

int DAC;                         /* File Handle for DSP */
unsigned int samplerate;         /* Samples/sec        */
unsigned int stereo;             /* stereo mono */
unsigned int afmt;               /* format for DSP  */
int Bufspersec;                  /* number of sounds fragments per second */
int Nfragbufs;                   /* number of driver frag buffers */
int fragsize;                    /* size of driver buffer fragments */
int fragsamplesize;              /* size of fragments in samples */
int LWn;                         /* number of specified loadable waveforms */
char **LWaa;                     /* array of specifed loadable waveforms */

    /* channel 1 the sweeping signal  ... */
char wf[32]="sawtooth";          /* waveform type */
unsigned int freq=1;             /* signal frequency */
int ratio=-1;                    /* used in pulse, sweep etc */
int Gain=1000;                   /* Amplification factor */

    /* channel 2  the swept signal ... */
char wf2[32]="sine";             /* waveform type */
unsigned int freqF=500;          /* signal low frequency */
unsigned int freqT=1500;         /* signal high frequency */
unsigned int freqC=1000;         /* signal centre frequency */
int frdev=500;                   /* signal deviation frequency */
int ratio2=-1;                   /* used in pulse, sweep etc */
int Gain2=1000;                  /* Amplification factor */
int phase=0;                     /* phase difference of chan2 from chan1 */

char *sys;
char *configfile;
char dac[130];

help(e)
int e;
{ 
  char **aa;
  
  fprintf(stderr,VERSTR,sys,VERSION);
  fputs("\nUsage: \n 1: sweepgen [flags] [waveform [freq [param]]]\n",stderr);
#ifdef HAVE_DAC_DEVICE
  fprintf(stderr,"Defaults: SINE wave swept by SAWTOOTH, output to %s, %d samples/sec,\n",
	  DAC_FILE,SAMPLERATE);
  fputs("         16 bit stereo samples if possible, else 8 bit and/or mono.\n",stderr);
#else
  fprintf(stderr,"Defaults: SINE wave swept by SAWTOOTH, %d samples/sec,\n",SAMPLERATE);
  fputs("         16 bit stereo samples. Must be used with -o or -w option.\n",stderr);
#endif

  fputs("Valid waveforms are:",stderr);
  for ( aa=(char **)getWavNames(); *aa; aa++ ) fprintf(stderr," %s",*aa);
  fputs("\n",stderr);
  fprintf(stderr,"Default Config files are \"%s\", \"$HOME/%s\" and\n\"%s\", searched in that order.\n",
          DEF_CONF_FILENAME, DEF_CONF_FILENAME, DEF_GLOBAL_CONF_FILE);

   fputs("flags: -s samples    generate with samplerate of samples/sec\n",stderr);
   fputs(" -8/-16 or -b 8|16   force 8 bit or 16 bit mode.\n",stderr);
   fputs("       -1    force mono - swept frequency output only\n",stderr);
   fputs("       -2    force stereo - swept freq on one chan, sweeping on other chan.\n",stderr);
   fputs("      -NB n  Number of Buffers to create is n, def. is 3\n",stderr);
   fputs("     -BPS n  Number of Buffers to play per sec, def. is 15\n",stderr);
/*  fputs("    -load wave_file  load waveform from wav_file\n",stderr);*/
   fputs("     -res n          n=1, 0.1 or 0.01  - frequency resolution\n",stderr);
   return(e);
}

/* main
 *
 */
 
main(argc,argv)
int argc;
char **argv;
{
   unsigned int v[MAXPRM],maxv,i,j,k,l,m,n,N;
   FILE *f;
   char *p,bf[130];
   int c;
   unsigned int t;
   
   if ((p=strrchr(sys=*argv++,'/'))!=NULL) sys=++p;
   argc--;
   
   configfile=DEF_CONF_FILENAME;
   samplerate=0; afmt=AFMT_QUERY;
   Bufspersec=0;
   Nfragbufs=0;
   vflg=dflg=0;
   stereo=-1;
   resolution=0;

   while (argc && **argv=='-') {          /* all flags and options must come */
      n=strlen(p=(*argv++)+1); argc--;    /* before paramters                */
      if (chkword(1,"samplerate")) {
	 if (argc && isdigit(**argv)) { samplerate=atoi(*argv++); argc--; }
      }
      else if (chkword(2,"NB")) {
	 if (argc && isdigit(**argv)) { Nfragbufs=atoi(*argv++); argc--; }
      }
      else if (chkword(3,"BPS")) { /* Buffers played per second - defines size */
	 if (argc && isdigit(**argv)) { Bufspersec=atoi(*argv++); argc--; }
      }
      else if (chkword(2,"16")) { afmt=AFMT_S16_LE; }
      else if (chkword(1,"bits")) {
	 i=0;
	 if (argc) { 
	    i=atoi(*argv++); argc--;
	 }
	 if (i==8) afmt=AFMT_U8;
	 else if (i==16) afmt=AFMT_S16_LE;
	 else exit(err_rpt(EINVAL,"must be '-b 8' or '-b 16'."));
      }
      else if (chkword(1,"A")) {
	 if (argc && isdigit(**argv)) { 
	    Gain=atoi(*argv++); argc--;
	 }
      }
      else if (chkword(3,"load")) {     /* load a waveform */
        if (argc) { LWaa[LWn++]=*argv++; argc--; } /* waveform file name */
      }
      else if (chkword(3,"resolution")) {  /* set freq generation resolution */
        if (argc) {
          if ((resolution=set_resolution(*argv++))==0) {
            exit(err_rpt(EINVAL,"bad resolution value, must be 1, 0.1 or 0.01"));
          }
          argc--;
        }
      }
      else {                              /* check for single char. flags    */
	 for (; *p; p++) {
	    if (*p=='h') exit(help(EFAIL));
	    else if (*p=='1') stereo=0;
	    else if (*p=='2') stereo=1;
	    else if (*p=='8') afmt=AFMT_U8;
	    else if (*p=='d') dflg=1;
	    else if (*p=='v') vflg++;
	    else {
	       *bf='-'; *(bf+1)=*p; *(bf+2)=0;
	       exit(help(err_rpt(EINVAL,bf)));
	    }
	 }
      }
   }
   
   if (argc) {
      strncpy(wf,*argv++,32); wf[31]=0; argc--;    /* waveform type */
      if (argc) {
	 freq=atoi(*argv++); argc--;
	 if (argc) { ratio=atoi(*argv++); argc--; }
	 if (argc) exit(help(err_rpt(EINVAL,"Too many parameters")));
      }
   }

   /* interrogate config file....... */

   init_conf_files(configfile,DEF_CONF_FILENAME,DEF_GLOBAL_CONF_FILE,vflg);
   if (vflg==0) {
      vflg=atoi(get_conf_value(sys,"verbose","0"));
   }
   if (samplerate==0) {
      samplerate=atoi(get_conf_value(sys,"samplerate",QSAMPLERATE));
   }
   if (stereo==-1) {
      stereo=atoi(get_conf_value(sys,"channels","1"));
      if (stereo) stereo--;
   }
   if (Nfragbufs==0) {
     Nfragbufs=atoi(get_conf_value(sys,"fragments",QDEFAULT_FRAGMENTS));
   }
   if (resolution==0) {
     resolution=set_resolution(get_conf_value(sys,"resolution",DEFAULT_RESOLUTION));
     if (resolution==0) {
       exit(err_rpt(EINVAL,"bad resolution value, must be 1, 0.1 or 0.01"));
     }
   }
   if (Bufspersec==0) {
     Bufspersec=atoi(get_conf_value(sys,"buffspersec",QDEFAULT_BUFFSPERSEC));
   }
   if (afmt==AFMT_QUERY) {
     afmt=atoi(get_conf_value(sys,"samplesize",QAFMT_QUERY));
   }
   strncpy(dac,get_conf_value(sys,"dacfile",DAC_FILE),sizeof(dac)-1);
   vikeys=atoi(get_conf_value(sys,"vi_keys",QVI_KEYS));

   /* if no format specified then try 16 bit */
   i=afmt;
   n=stereo;
   if ((DAC=DACopen(dac,"w",&samplerate,&i,&n))<0) {
      exit(err_rpt(errno,dac));
   }
   if (((afmt!=AFMT_QUERY) && (i!=afmt)) ||
       ((stereo!=-1) && (n!=stereo))) {
      exit(err_rpt(EINVAL,"Sound card doesn't support format requested."));
   }
   afmt=i; stereo=n;

   if ((fragsize=setfragsize(DAC,Nfragbufs,Bufspersec,samplerate,afmt,stereo))<0) {
      exit(err_rpt(errno,"Problem setting appropriate fragment size."));
   }
   fragsamplesize=(fragsize>>(afmt==AFMT_S16_LE))>>(stereo);
       
   if (freq > samplerate/2) {
      fprintf(stderr,"%d Hz is more than half the sampling rate\n",freq);
   }

   if (vflg) {
      printf("%s %s bit samples being generated.\n",
	     (stereo)?"Stereo, ":"Mono, ",(afmt==AFMT_S16_LE)?"16":"8");
      printf("Samples scaled by a factor of %d/1000\n",Gain);
      printf("Playing at %d samples/sec\n",samplerate);
      printf("Frequency resolution is %.2fHz\n",1.0/resolution);
      printf("Buffer fragment size is %d bytes (%d samples). Aprox. %d millisecs.\n",
	     fragsize, fragsamplesize,
	     1000*((fragsize>>(stereo))>>(afmt==AFMT_S16_LE))/samplerate);
      printf("Requested %d buffers/sec and have %d buffs/sec\n",Bufspersec,
	     (samplerate+(fragsamplesize/2))/fragsamplesize);
   }

   WinGen(DAC);
   
   close(DAC);
   exit(0);
}  


set_resolution(p)
char *p;
{
  char *s;
  int r;
  double d;

  d=strtod(p,&s); r=0;
  if (d==1.0) r=1;
  else if (d==0.1) r=10;
  else if (d==0.01) r=100;
  return(r);
}

